### DISCUSSION TITLE
RAG API taking extremely long to process documents over 1mb?

### DISCUSSION THREAD

**Discussion by ggomp2885** [danny-avila/LibreChat#4081] [Troubleshooting] [answered]
### What happened?

When I upload a 1.3mb pdf file as a test, Its taking over 1 hour or more to get into the rag database. And its not letting me cancel the upload once it starts, so the file is just stuck sitting there for 1 hour, blocking any other file from being uploaded in the UI.
I have to force-kill the enitre LibreChat app and restart it to have this reset.

This makes uploading any kind of decent length document, just not usable at all,
is there a setting that I am missing?

### Steps to Reproduce

1. Open LibreChat
2. Set embedding model to: Azure, "text-embedding-3-small"
3. Attempt to upload a 1mb or larger pdf file into the rag_api

### What browsers are you seeing the problem on?

Chrome

### Relevant log output

```shell
2024-09-12 19:17:35,652 - multipart.multipart - DEBUG - Calling on_part_data with data[0:1]
2024-09-12 19:17:35,653 - multipart.multipart - DEBUG - Calling on part_data with data[48218:130477]
2024-09-12 19:17:35,655 - multipart.multipart - DEBUG - Calling on part_data with data[0:10103]
2024-09-12 19:17:35,655 - multipart.multipart - DEBUG - Calling on_part_data with data[0:1]
2024-09-12 19:17:35,655 - multipart.multipart - DEBUG - Calling on part_data with data[10104:36314]
2024-09-12 19:17:35,656 - multipart.multipart - DEBUG - Calling on part_data with data[0:1]
```

### Screenshots

![image](https://github.com/user-attachments/assets/3eb78f03-1a14-4158-8aa4-7f4ef994a822)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by danny-avila** [comment]
This is not the expected experience at all, I can upload 10+ mb PDFs rather quickly. Really the main thing that affects upload time is the amount of text content from the files, no matter what size or file type it is. 

Is this a text "heavy" PDF you are trying to upload? Are there any other files that work for you?

First be sure everything else is configured right, too: https://www.librechat.ai/docs/configuration/rag_api

**Reply by ggomp2885** [reply]
Hey Danny, 
Thanks so much for your response!
Here is the example we are working with, 

Even with just the first 5 pages of this document, It is taking 3min 30 seconds to get into the rag_api,
Example here: [first_5_pages.pdf](https://github.com/user-attachments/files/17050778/first_5_pages.pdf)

Whereas, Chatgpt.com (as a reference point) can upload the same 5 pages in 1 second.

Attaching my both my LibreChat .env file and my rag_api .env file here,
[librechat_env.txt](https://github.com/user-attachments/files/17050620/librechat_env.txt)
[rag_api_env.txt](https://github.com/user-attachments/files/17050624/rag_api_env.txt)

We are running this locally on a very large server, using the "npm run backend" command,
Would you expect this to make any difference in the rag_api speed versus running it as a docker container?
Is 3min 30sec for 5 pages of text the expected experience?

**Reply by danny-avila** [reply]
No it's not expected at all, watch my video, it takes approx. 1 second as well with the pdf you attached. No clue what the bottleneck is for you


https://github.com/user-attachments/assets/001b0c8b-0d81-4a8f-8f10-4ac98fa129c1

**Reply by ggomp2885** [reply, marked as answer]
Hey Danny,
Good news!!
Found out the source of this issue,
It was the "PDF_EXTRACT_IMAGES" parameter in our .env file, this needs to be set to false,

This parameter gets read into the "extract_images" parameter in the pypdfloader library,

When this was set to True, it would take 45seconds or more per page to get through the loader.load() function in main.py in rag_api repo. When this is set to False, it completes this function in half a second or less.

![MicrosoftTeams-image](https://github.com/user-attachments/assets/10944621-4c73-4492-adce-fcd3b44f4ff1)

Thank you so much for posting this video of you uploading it on yourside,
This was exactly the confirmation we needed to know it was definitely coming from our side!

**Comment by graycyanai** [comment]
Interesting issue—this kind of delay can seriously impact workflows, especially when scaling document ingestion. The root cause with the PDF_EXTRACT_IMAGES parameter makes sense, since image extraction can significantly slow down processing.

For anyone working on [RAG AI Manufacturing](https://graycyan.ai/rag-ai-manufacturing/) use cases, optimizing ingestion pipelines like this is critical. Disabling unnecessary image parsing (when not needed) seems like a smart move to keep latency low and improve overall system efficiency.

Appreciate the detailed debugging and solution shared here—super helpful!
